### ISSUE TITLE
Sync-Catalog: Configure health-check 

### ISSUE THREAD

**Issue by sagarj8n** [hashicorp/consul-k8s#29]
How do you configure the health check for the service that is automatically registered using the Consul sync Catalog.?

**Comment by madsonic** [comment]
strangely even the demo shows a dash for consul-ui
https://www.hashicorp.com/blog/consul-and-kubernetes-service-catalog-sync
I have some services that automatically shows passing health check while some showing merely a dash (as per the demo). Is there anything to do in particular to show health checks for synced services in kubernetes?

**Comment by arnodenuijl** [comment]
Just to chime in. I have a case where i need a health check here too. 

We are running a setup with consul and fabio (https://fabiolb.net/). Fabio is a reverse proxy that automatically (via consul tags) registers consul servers in the reverse proxy. 

I have consul-k8s setup via helm and when i deploy a service in kubernetes I see the service appearing in consul including the 'urlprefix-' tag (that fabio uses) that I added. The only problem is that fabio only registers services that have a health check. So since consul-k8s doesn't add an explicit health check the service doesn't appear in the reverse proxy.

```
---
apiVersion: v1
kind: Service
metadata:
  name: mynginx
  annotations:
    consul.hashicorp.com/service-tags: "urlprefix-/mynginx"
    consul.hashicorp.com/service-sync: "true"
    consul.hashicorp.com/service-name: "mynginx"
  labels:
    run: mynginx
spec:
  ports:
  - port: 80
    protocol: TCP
  selector:
    run: mynginx
```

**Comment by tasdikrahman** [comment]
@sagarj8n @madsonic @arnodenuijl did you folks figure it out?  Landed here on a google search for an answer for the same.

**Comment by arnodenuijl** [comment]
No. I must say I didn't spend a lot of time on the issue anymore.  But I didn't see a way to get this working.

**Comment by jensoncs** [comment]
@adilyse @mitchellh Can you confirm whether Consul sync catalog is providing any health check or not? If not is there any plan to add the functionality?

**Comment by yarinm** [comment]
+1 for this
I also encountered a scenario that I'd like to have a health check defined in consul. I might have to resort to manually sync the service and define the health checks. But this should be solved in catalog sync.

**Comment by dfang** [comment]
k8s have health check，maybe catalog sync should sync status from that

**Comment by gmachine1** [comment]
I'm looking for this too looks like it's not there, so instead will manually add the desired health checks.

**Comment by gmachine1** [comment]
in particular what would be ideal in my case is add some annotations such as 
```
"consul.hashicorp.com/service-meta-check_tcp": "true"                                              
"consul.hashicorp.com/service-meta-check_interval": "5s"
``` 
(as was done when using registrator and docker-compose) and automatically a tcp check is added

**Comment by mstrYoda** [comment]
Any update here? We are now using consul catalog sync to sync service from k8s to consul. We want to add health checks to consul for services. I want to add such a feature adds health check via custom annotations like @gmachine1 said. It seems it is enough to check tcp connection to <service-ip>:<node-port>.

Any thoughts @lkysow ?

**Comment by KalenWessel** [comment]
Hoping for an update on this? I'm starting to use connect-inject which does a lot of things behind the scenes via annotations but still doesn't support service health checks which seems pretty important.

**Comment by mstrYoda** [comment]
I am volunteer to make a pr for this. We are using consul catalog snyc and we need to add another health check layer to our application. Instead consul catalog can handle this.

We are registering services from k8s and vms to consul. In some cases, we need to health check that ip addresses. I will update my comment when I remember the exact case.

**Comment by david-yu** [comment]
Hi there, Consul PM here. It sounds like there are two distinct ideas on how one would want to consume Healthchecks. One is to configure an actual Consul Health check on Kubernetes, while another is to sync a Kubernetes health check with Consul to help direct traffic to healthy service instances. 

We are currently leaning towards the latter since it aligns better with Kubernetes design and would not require explicit annotations to define any health checks. I'm curious to know whether those that asked for native Consul health checks whether defining those health checks as Kubernetes readiness probes would be a workable solution, which Consul Kubernetes can then pick up to direct traffic to healthy service instances.

**Comment by filintod** [comment]
@david-yu I also want the latter (sync kubernetes health checks with Consul. Thanks

**Comment by JasonGilholme** [comment]
Yes, the latter please!

**Comment by raypettersen** [comment]
@david-yu what's the status on this feature?

**Comment by phr0zenx** [comment]
@david-yu Is there a way I can implement health checks via k8s to check on the services via consul? I want to implement pod regenerative triggers that k8s is able to execute. Unless consul can do that?

**Comment by lkysow** [comment]
Hi Folks, I'm curious what types of services you're syncing to Consul.

NodePort and LoadBalancer services don't necessarily make sense to have their health synced to Consul because when the request lands at the node port or load balancer, Kubernetes handles routing to only healthy pods.

If you're syncing ClusterIP services, then the individual pod ips are synced to Consul and in this case it does make sense to sync the readiness status of each pod.

**Comment by ak08743** [comment]
Hi @david-yu 

Are there any updates on this issue ?

**Comment by ndhanushkodi** [comment]
@ak08743 we now do syncing of Kubernetes readiness status so Consul's health check state will reflect that. Docs are [here](https://www.consul.io/docs/k8s/connect/health).

**Comment by ak08743** [comment]
@ndhanushkodi Thank you for a quick reply. 

The document that you linked is about service mesh and Consul cluster installed in k8s environment, unless I am misunderstanding something. This issue, and the problem we are experiencing now, is that we have external Consul cluster, and use ServiceSync to expose selected k8s services into that external Consul cluster for 3rd-party consumption. Reading through the linked document, I do not see how it would help in this case.

**Comment by david-yu** [comment]
Hi @ak08743 and folks following along. This issue is still outstanding. We initially had thought we could address this as part of a larger effort to build in health check support for Consul on Kubernetes, however we had to move this to a future release due to scope. 

As pointed by @lkysow could folks chime in on what kind of Services you are syncing with Consul K8s and how you expect the health to be represented if one of multiple pods for a service goes down? As an example in the case of LoadBalancer and NodePort services if one of the pods is unhealthy, the service may still be considered healthy since K8s would automatically route traffic to health pods.

**Comment by ruaridhangus-fbr** [comment]
In our case we have an inter-cluster issue EKS and Bare Metal with 2 different types of service however one type is being removed as its legacy.

1) EKS service is ClusterIP synced to Consul (AWS ENI IP address of the pod to be clear) which is then accessible to the Bare Metal cluster.
2) NodePort service being synced to Consul with hardcoded port (this is a legacy hangover to us and being deprecated so not the end of the world if pod is not healthchecked).

In regards to what is healthy for us is if there is 1 or more healthy pod in the service then the service can be classed as healthy but the individual pod is unhealthy and shouldn't be included in DNS lookup.

**Comment by ak08743** [comment]
In our use case, we expose only LoadBalancer services outside the k8s cluster, so as long as there is something, alive and healthy, that can process the request - the service as a whole should be marked as healthy.

**Comment by webmutation** [comment]
Just adding to the mix that a different consul sync project has added support for health checks
https://github.com/lynes-io/kube-consul-register
In this case they are using an annotation.

To have a working Consul+FabioLB setup it would be enough if the Virtual node k8s-sync had a health check. Then we could do routing based on that value. It only needs one passing health check node or service, not both.

**Comment by david-yu** [comment]
Hi @webmutation sorry for the delay here. We are open to PRs to address and can review them to address changes  in Catalog Sync. Although we still do maintain Catalog sync, we have arrived at a state where we don't anticipate any more changes to the product since our focus has shifted to building a service mesh product that supersedes the original work that was built into Catalog Sync, by providing service discovery across both K8s and VM clusters through  the mesh. This may not be what you were hoping for, but I wanted to communicate our current stance and thinking on this issue.

**Comment by Juandavi1** [comment]
in short, do not recommend using this tool ? healths management is critical and we can't use consul as a service mesh because we already use istio

**Comment by david-yu** [comment]
Closing out as this is addressed by https://github.com/hashicorp/consul-k8s/pull/1821. This will be released with Consul K8s 1.0.x and Consul K8s 1.1.x.
